Add RavenDB client & hosting integration

shiranshalom opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

<!-- ⚠️ We will not merge the PR to the main repo if your changes are not in a *feature branch* of your forked repository. Create a new branch in your repo, **do not use the `main` branch in your repo**! -->

<!-- ⚠️ **Every PR** needs to have a linked issue and have previously been approved. PRs that don't follow this will be rejected. >

<!-- 👉 It is imperative to resolve ONE ISSUE PER PR and avoid making multiple changes unless the changes interrelate with each other -->

<!-- 📝 Please always keep the "☑️ Allow edits by maintainers" button checked in the Pull Request Template as it increases collaboration with the Toolkit maintainers by permitting commits to your PR branch (only) created from your fork. This can let us quickly make fixes for minor typos or forgotten StyleCop issues during review without needing to wait on you doing extra work. Let us help you help us! 🎉 -->

**Closes #[396](https://github.com/CommunityToolkit/Aspire/issues/396)**

<!-- Add an overview of the changes here -->

<!-- All details should be in the linked issue. Feel free to call out any outstanding differences here. -->

This pull request includes several updates to the `Aspire.CommunityToolkit.sln` file, new project additions, and updates to dependencies. The most important changes include the addition of new RavenDB projects, updates to project references, and the introduction of new dependencies.

#### Project Additions and Updates:
Added new RavenDB projects to the solution, including `Aspire.CommunityToolkit.Hosting.RavenDB`, `Aspire.CommunityToolkit.RavenDB.Client`, and their respective test and example projects .

#### Dependency Updates:
Added new dependencies in `Directory.Packages.props`, including `RavenDB.Client`, `AspNetCore.HealthChecks.RavenDB`, and `RavenDB.TestDriver`.

## PR Checklist

- [x] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [x] Based off latest main branch of toolkit
- [x] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [x] New integration
  - [ ] Docs are written - this PR includes `README.md` files for RavenDB client and hosting projects. Complete documentation is prepared, and we will open a PR to the [dotnet/docs-aspire](https://github.com/dotnet/docs-aspire) repository soon.
  - [ ] Added description of major feature to project description for NuGet package (4000 total character limit, so don't push entire description over that)
- [x] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [x] Every new API (including internal ones) has full XML docs
- [x] Code follows all style conventions

## Other information

We are currently working on an article related to this integration. Once it's ready, I will attach the link here.
Additionally, some links in this PR will need to be updated once the NuGet packages are published.

shiranshalom commented:
`@dotnet-policy-service agree company="Hibernating Rhinos"`

aaronpowell commented on tests/CommunityToolkit.Aspire.RavenDB.Client.Tests/AspireRavenDBExtensionsTests.cs:
```diff
@@ -0,0 +1,410 @@
+        Assert.Equal(DefaultConnectionString, documentStore1.Urls[0]);
+        Assert.Equal(DefaultConnectionString, documentStore2.Urls[0]);
+
+        Assert.True(documentStore1.Database == databaseName1 || documentStore1.Database == databaseName2);
+        Assert.True(documentStore2.Database == databaseName1 || documentStore2.Database == databaseName2);
```
Why would the store be one of two databases? Shouldn't that be a known assert?

shiranshalom replied:
Removed

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/HealthChecksExtensions.cs:
```diff
@@ -0,0 +1,66 @@
+    {
+        return builder.Add(new HealthCheckRegistration(
+            name ?? NAME,
+            sp => new RavenDBHealthCheck(new RavenDBOptions() { Urls = new[] { connectionStringFactory(sp) } }),
```
Changed as requested.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/HealthChecksExtensions.cs:
```diff
@@ -0,0 +1,66 @@
+    {
+        return builder.Add(new HealthCheckRegistration(
+            name ?? NAME,
+            sp => new RavenDBHealthCheck(new RavenDBOptions() { Urls = new[] { connectionStringFactory(sp) }, Database = databaseName }),
```
Changed as requested.

shiranshalom commented on tests/CommunityToolkit.Aspire.RavenDB.Client.Tests/AspireRavenDBExtensionsTests.cs:
```diff
@@ -0,0 +1,410 @@
+    private string DefaultConnectionString => _serverFixture.ConnectionString ??
+                                              throw new InvalidOperationException("The server was not initialized.");
+
+    public AspireRavenDBExtensionsTests(RavenDbServerFixture serverFixture)
```
Done.

shiranshalom commented on tests/CommunityToolkit.Aspire.Hosting.RavenDB.Tests/AppHostTests.cs:
```diff
@@ -0,0 +1,61 @@
+        Assert.Equal(connectionString.OriginalString, url);
+        Assert.Equal("TestDatabase", dbResource.DatabaseName);
+
+        await Task.Delay(10000, cancellationToken.Token);
```
Sorry, it shouldn't be there. Removed.

shiranshalom commented on src/CommunityToolkit.Aspire.RavenDB.Client/RavenDBClientExtension.cs:
```diff
@@ -0,0 +1,264 @@
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        if (settings.Urls == null || settings.Urls.Length == 0)
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.RavenDB.Client/RavenDBClientExtension.cs:
```diff
@@ -0,0 +1,264 @@
+        X509Certificate2? certificate = null)
+    {
+        var settings = new RavenDBClientSettings(connectionUrls, databaseName);
+        if (certificate != null)
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.RavenDB.Client/RavenDBClientExtension.cs:
```diff
@@ -0,0 +1,264 @@
+        X509Certificate2? certificate = null)
+    {
+        var settings = new RavenDBClientSettings(connectionUrls, databaseName);
+        if (certificate != null)
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.RavenDB.Client/CommunityToolkit.Aspire.RavenDB.Client.csproj:
```diff
@@ -0,0 +1,16 @@
+    <AdditionalPackageTags>client ravendb</AdditionalPackageTags>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
```
Fixed.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/RavenDBServerSettings.cs:
```diff
@@ -0,0 +1,151 @@
+    public string PublicServerUrl { get; set; }
+}
+
+#pragma warning disable RS0016
```
Removed :)

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/RavenDBServerSettings.cs:
```diff
@@ -0,0 +1,151 @@
+﻿namespace CommunityToolkit.Aspire.Hosting.RavenDB;
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring as nullable.
```
Removed.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/RavenDBServerResource.cs:
```diff
@@ -0,0 +1,58 @@
+    /// </summary>
+    /// <param name="name">The name of the RavenDB server resource.</param>
+    /// <param name="isSecured">Indicates whether the server connection is secured (true for HTTPS, false for HTTP).</param>
+    public RavenDBServerResource(string name, bool isSecured) : base(name)
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/RavenDBBuilderExtensions.cs:
```diff
@@ -0,0 +1,219 @@
+        {
+            connectionString = await databaseResource.ConnectionStringExpression.GetValueAsync(ct).ConfigureAwait(false);
+
+            if (connectionString == null)
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/RavenDBBuilderExtensions.cs:
```diff
@@ -0,0 +1,219 @@
+            environmentVariables.TryAdd("RAVEN_PublicServerUrl", securedServerSettings.PublicServerUrl);
+            environmentVariables.TryAdd("RAVEN_Security_Certificate_Path", securedServerSettings.CertificatePath);
+            
+            if (securedServerSettings.CertificatePassword != null)
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/RavenDBBuilderExtensions.cs:
```diff
@@ -0,0 +1,219 @@
+            environmentVariables.TryAdd("RAVEN_License", serverSettings.LicensingOptions.License);
+        }
+
+        if (serverSettings.ServerUrl != null)
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/RavenDBBuilderExtensions.cs:
```diff
@@ -0,0 +1,219 @@
+            { "RAVEN_Setup_Mode", serverSettings.SetupMode.ToString() }
+        };
+
+        if (serverSettings.LicensingOptions != null)
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/RavenDBBuilderExtensions.cs:
```diff
@@ -0,0 +1,219 @@
+        {
+            connectionString = await serverResource.ConnectionStringExpression.GetValueAsync(ct).ConfigureAwait(false);
+
+            if (connectionString == null)
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/README.md:
```diff
@@ -0,0 +1,31 @@
+## Additional documentation
+
+<!-- TODO: Update the link once it is created -->
+https://learn.microsoft.com/dotnet/aspire/community-toolkit/hosting-ravendb
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/CommunityToolkit.Aspire.Hosting.RavenDB.csproj:
```diff
@@ -0,0 +1,19 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Include="..\Shared\VolumeNameGenerator.cs" Link="Utils\VolumeNameGenerator.cs" />
```
Done.

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/CommunityToolkit.Aspire.Hosting.RavenDB.csproj:
```diff
@@ -0,0 +1,19 @@
+    <AdditionalPackageTags>hosting ravendb</AdditionalPackageTags>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
```
Done.

shiranshalom commented on Directory.Packages.props:
```diff
@@ -23,6 +23,9 @@
     <PackageVersion Include="Swashbuckle.AspNetCore" Version="7.2.0" />
+    <PackageVersion Include="RavenDB.Client" Version="6.2.1" />
+    <PackageVersion Include="RavenDB.TestDriver" Version="6.2.1" />
+    <PackageVersion Include="AspNetCore.HealthChecks.RavenDB" Version="9.0.0" />
```
Done.

shiranshalom commented on CommunityToolkit.Aspire.sln:
```diff
@@ -221,6 +221,18 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CommunityToolkit.Aspire.Mic
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "ravendb", "ravendb", "{1C65F967-D5F1-424B-82E9-B8585B6F0BD6}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "RavenDB.AppHost", "examples\ravendb\RavenDB.AppHost\RavenDB.AppHost.csproj", "{28FCB1E2-7460-4FDD-AA63-03162C9F2154}"
```
Right! Sorry, done.

shiranshalom commented:
@aaronpowell thanks for the feedback!

I’ll work on aligning the client integration and I’ll push these updates shortly.

shiranshalom commented on src/CommunityToolkit.Aspire.RavenDB.Client/RavenDBClientExtension.cs:
Done :)

shiranshalom commented on src/CommunityToolkit.Aspire.Hosting.RavenDB/RavenDBDatabaseResource.cs:
```diff
@@ -0,0 +1,40 @@
+    /// <summary>
+    /// Gets the connection string expression for the RavenDB database, derived from the parent server's connection string.
+    /// </summary>
+    public ReferenceExpression ConnectionStringExpression => Parent.ConnectionStringExpression;
```
Done

shiranshalom commented:
@aaronpowell 
I’ve completed the changes you requested. Here’s a summary of the updates:

Client Integration:
* Updated the integration to fully support dependency injection and added support for configuration using a connection string. We opted to keep some of the extra methods to allow clients to use the library independently of the hosting integration.

Connection String Format:
* Modified the `ConnectionStringExpression` to the format: `URL={Url};Database={DatabaseName}`. We chose `URL` instead of `ENDPOINT` because we believe it aligns better with our code conventions and feels more intuitive.

Expanded Example Code:
* Expanded the example in the codebase to include a new `ApiService` project and added some basic API endpoints to demonstrate practical use cases of the integration.

Tests:
* Fixed existing tests to align with the new integration changes.
* Added new tests to cover the updated functionality and improve overall test coverage.

If you have any additional feedback or suggestions, I’d be happy to address them!

aaronpowell commented:
Sorry for the delay, I'll hope to review the updates this week

aaronpowell commented:
Last two things @shiranshalom 

1. Can you update the CODEOWNERS file with you as the owner.
2. Can you update the README with the packages and their details (don't worry about broken links).

Regarding docs, do you want them to be on the RavenDB docs, or is the intent to be on the Microsoft Learn docs?

shiranshalom commented:
> Last two things @shiranshalom
> 
> 1. Can you update the CODEOWNERS file with you as the owner.

Done

> 2. Can you update the README with the packages and their details (don't worry about broken links).

Done

> Regarding docs, do you want them to be on the RavenDB docs, or is the intent to be on the Microsoft Learn docs?

Both! 🙂 I'll be submitting a new PR to the [dotnet/docs-aspire](https://github.com/dotnet/docs-aspire) repository, and there will also be documentation and articles on our website. I'll share the links as soon as they're available

aaronpowell reviewed (approved):
All good @shiranshalom and I've sent an invite to join the contributors (I thought I'd already done that, whoops!)
